openhtf/io/frontend/__main__.py

```diff
@@ -32,97 +32,207 @@
-  """Query a running instance of OpenHTF for state."""
-  result = None
-  data = cache()[station_id]
-  conn = httplib.HTTPConnection('localhost', data.http_port)
-  args = [method, '/']
-  if message:
-    args.append(message)
-  try:
-    conn.request(*args)
-    if method == 'GET':
-      response = conn.getresponse()
-      if response.status != 200:
-        result = {'state': 'BORKED'}
-      else:
-        result = json.loads(response.read())
-  except socket.error as e:
-    result = {'state': 'OFFLINE'}
-  conn.close()
-  return result
-
-
-@app.route('/', methods=['GET'])
-def station_list():
-  """Handle a request for a list of known local OpenHTF stations."""
-  stations = [
-      {'data': data,'status': 'ONLINE' if data.IsAlive() else 'OFFLINE'}
-      for _, data in cache().iteritems()]
-  return flask.render_template('station_list.html',
-                               stations=stations,
-                               host=socket.gethostname())
-
-
-@app.route('/station/<station_id>/', methods=['GET'])
-def station(station_id):
-  """Handle a request for the state of a particular station."""
-  state = query_framework(station_id)
-  return flask.render_template('station.html', name=station_id, state=state)
-
-
-@app.route('/station/<station_id>/template/<template>/')
-def station_template(station_id, template):
-  """Handle requests for individual pieces of station pages."""
-  state = state = query_framework(station_id)
-  return flask.render_template(
-      '%s.html' % template, name=station_id, state=state)
-
-
-@app.route('/station/<name>/prompt/', methods=['GET'])
-def get_prompt(name):
-  """Handle a request for a station's prompt state."""
-  state = query_framework(name)
-  print('HACK 1')
-  if ('prompt' not in state) or (state['prompt'] == 'None'):
-    return 'NO_PROMPT'
-  return state['prompt']['id']
-
-
-@app.route('/station/<name>/prompt/<id>/', methods=['POST'])
-def prompt(name, id):
-  """Handle requests that respond to prompts."""
-  msg = json.JSONEncoder().encode({'id': id, 'response': flask.request.data})
-  query_framework(name, method='POST', message=msg)
-  return 'OK'
+gflags.DEFINE_integer('discovery_interval_s', 5,
+                      'Seconds to wait between station discovery attempts.')
+gflags.DEFINE_string('discovery_address',
+                     multicast.DEFAULT_ADDRESS,
+                     'Multicast address to ping for station discovery.')
+gflags.DEFINE_integer('discovery_port',
+                      multicast.DEFAULT_PORT,
+                      'Multicast port to use for station discovery.')
+gflags.DEFINE_integer('discovery_ttl',
+                      multicast.DEFAULT_TTL,
+                      'TTL for station discovery pings.')
+
+
+_LOG = logging.getLogger(__name__)
+
+UNKNOWN_STATION_ID = 'UNKNOWN_STATION'
+
+
+class Station(mutablerecords.Record(  # pylint: disable=too-few-public-methods,no-init
+    'Station', [], {'station_id': UNKNOWN_STATION_ID, 'state': None})):
```

Comment:
might want to track the last update time in here too and periodically re-update the state of stale stations just for kicks.